Add context to all requests

obitech opened the pull request:
This adds the possibility to provide a `context.Context` to all GET
and POST requests and introduces:

- `LookupWithContext` calls `RawLookupWithContext` calls `newGetRequestWithContext`
- `BulkLookupWithContext` calls `RawBulkLookupWithContext` calls `newPostRequestWithContext`

The new functions are duplicated from the old ones but use a `NewRequestWithContext` instead.
The old functions call the `WithContext` functions but provide a `context.Background()`, so
only one set needs to be maintained.

Closes #14

theckman commented on client.go:
```diff
@@ -215,7 +240,7 @@ func newBulkPostRequest(urlStr, apiKey string, ips []string) (*http.Request, err
 		return nil, errors.Wrap(err, "failed to encode JSON")
```
Now that we have the `ctx`, we should check `if ctx.Err() != nil` before marshaling this JSON. If the context is already canceled no use doing this work.

Are there other places we should check for cancelation now to avoid unnecessary work?

costela replied:
Isn't this a bit of premature optimization? JSON _encoding_ isn't quite as bad as _decoding_.

obitech replied:
I've added an extra check, looks like it's the only place worth checking 🙂

theckman replied:
@costela _Potentially_. Both encoding and decoding make use of reflection, and it'll have allocated memory for the JSON output that we're discarding. It doesn't seem unreasonable to avoid both cases if we know we're going to throw it all away.

theckman commented on client.go:
```diff
@@ -210,12 +235,21 @@ func newBulkPostRequest(urlStr, apiKey string, ips []string) (*http.Request, err
+		case context.Canceled:
+			return nil, errors.New("context cancelled")
+		}
+	}
```
I would simplify the whole block to:

```Go
if err := ctx.Err(); err != nil {
  return nil, err
}
```

I think you'll even end up with the same error messages. 😄

obitech replied:
No worries, I fixed it.

obitech commented:
> What are your thoughts on that?

Good idea! I deleted the function and adjusted the tests. I can't run `make tests` locally though since there seems to be some service expected to be running on `:9085`? Maybe we can add a little note to `README.md` describing what contributors need to setup to run the tests locally 🙂

obitech commented:
Hey @theckman you have any feedback on my latest changes? 🙂

jonathan-kosgei commented:
@theckman is this good to go?

obitech commented:
Thank you for merging this @jonathan-kosgei 🎉 

Can we get a new release with this change? 🙂

jonathan-kosgei commented:
@obitech Done! Apologies for the delay and thank you for your contributing! :smile:

obitech commented:
Thanks a lot @jonathan-kosgei ! 🙏
